### ISSUE TITLE
reduce source scientificname without speciesannex and solve duplicates

### ISSUE THREAD

**Issue by damianooldoni** [inbo/speciesbim#41]
There are duplicates in new scientificname table. This has been documented here: https://github.com/inbo/speciesbim/issues/36#issuecomment-690952138 but it is worth a new issue to be better tackled.

This a copy paste of the table with duplicates mentioned in that post:

id | deprecatedTaxonId | scientificName | authorhsip
-- | -- | -- | --
4859 | 26718 | Agaricus sylvicola | (Vittad.) Peck, 1872
6818 | 50129 | Agaricus sylvicola | 
7743 | 51280 | Agonum emarginatum | (Gyllenhal, 1827)
2169 | 16405 | Agonum emarginatum | (Gyllenhal, 1827)
4871 | 26780 | Amanita excelsa | (Fr.:Fr.) Bertill.
7945 | 51486 | Amanita excelsa | Bertill., 1866 (Fr.)
8252 | 51819 | Amanita excelsa | Bertill., 1866 (Fr.)
8138 | 51681 | Anas clypeata | Linnaeus, 1758
720 | 3521 | Anas clypeata | Linnaeus, 1758
713 | 3510 | Anas penelope | Linnaeus, 1758
8146 | 51689 | Anas penelope | Linnaeus, 1758
8156 | 51699 | Anas querquedula | Linnaeus, 1758
715 | 3512 | Anas querquedula | Linnaeus, 1758
8140 | 51683 | Anas sibilatrix | Poeppig, 1829
6842 | 50153 | Anas sibilatrix | Poeppig, 1829
716 | 3513 | Anas strepera | Linnaeus, 1758
8139 | 51682 | Anas strepera | Linnaeus, 1758
2956 | 20772 | Anchusa arvensis | (L.) M.Bieb.
8012 | 51554 | Anchusa arvensis | M. Bieb. (L.)
2527 | 18127 | Anoplotrupes stercorosus | (Scriba, 1791)
8154 | 51697 | Anoplotrupes stercorosus | (Scriba, 1791)
6874 | 50185 | Arabidopsis arenosa subsp. arenosa | (L.) Lawalrée
6873 | 50184 | Arabidopsis arenosa subsp. arenosa | (L.) Lawalrée
6877 | 50188 | Arctium minus | (Hill) Bernh.
2882 | 20672 | Arctium minus | (Hill) Bernh.
5831 | 34306 | Bombus | Latreille, 1802
7718 | 51039 | Bombus | Latreille, 1802
7888 | 51429 | Bromus madritensis | L.
7961 | 51502 | Bromus madritensis | L.
6944 | 50252 | Calamagrostis epigejos | (L.) Roth
3301 | 21247 | Calamagrostis epigejos | (L.) Roth
8285 | 51853 | Caloplaca aurantia | Hellb. (Pers.)
5796 | 34158 | Caloplaca aurantia | (Pers.) Hellbom
8288 | 51856 | Caloplaca decipiens | Blomb. & Forssell (Arnold)
5799 | 34165 | Caloplaca decipiens | (Arnold) Blomb. & Forssell
8293 | 51861 | Caloplaca flavescens | J.R.Laundon (Huds.)
5800 | 34166 | Caloplaca flavescens | (Hudson) J.R. Laundon
5802 | 34169 | Caloplaca flavovirescens | (Wulfen) Dalla Torre & Sarnth.
8304 | 51872 | Caloplaca flavovirescens | Dalla Torre & Sarnth. (Wulfen)
8074 | 51616 | Cardamine flexuosa | With.
2979 | 20801 | Cardamine flexuosa | With.
7663 | 50980 | Carduelis flammea | (Linnaeus, 1758)
698 | 3484 | Carduelis flammea | (Linnaeus, 1758)
4949 | 27588 | Clitocybe gibba | (Pers.:Fr.) P. Kumm.
8256 | 51823 | Clitocybe gibba | P.Kumm. (Pers.)
7028 | 50333 | Cortinarius livido-ochraceus | (Berk.) Berk. 1860
986 | 4624 | Cortinarius livido-ochraceus | (Berk.) Berk.
8124 | 51666 | Craspedosoma rawlinsii | Leach, 1815
1233 | 5704 | Craspedosoma rawlinsii | Leach, 1814
3411 | 21410 | Crataegus laevigata | (Poir.) DC.
7773 | 51313 | Crataegus laevigata | DC. (Poir.)
7754 | 51293 | Cryphia domestica | Hufnagel, 1766
4011 | 22818 | Cryphia domestica | (Hüfnagel, 1766)
601 | 3264 | Dactylorhiza fuchsii | (Druce) Soó
7731 | 51054 | Dactylorhiza fuchsii | Soó (Druce)
667 | 3435 | Dendrocopos medius | (Linnaeus, 1758)
8141 | 51684 | Dendrocopos medius | (Linnaeus, 1758)
8142 | 51685 | Dendrocopos minor | (Linnaeus, 1758)
668 | 3436 | Dendrocopos minor | (Linnaeus, 1758)
7079 | 50385 | Draba verna | L.
7785 | 51325 | Draba verna | L.
7089 | 50395 | Enoplognatha | 
6457 | 41327 | Enoplognatha | 
7143 | 50450 | Galerina vittiformis | (Fr.) Singer 1950
8241 | 51784 | Galerina vittiformis | Singer, 1950 (Fr.)
8011 | 51553 | Galium parisiense | L.
7794 | 51334 | Galium parisiense | L.
5177 | 29438 | Hemimycena lactea | (Pers.:Fr.) Singer
8258 | 51825 | Hemimycena lactea | Singer (Pers.)
8242 | 51785 | Hygrocybe conica | P.Kumm. (Schaeff.)
5205 | 29642 | Hygrocybe conica | (Schaeff.:Fr.) P. Kumm.
1048 | 4988 | Hygrocybe glutinipes | (J.E. Lange) R. Haller
8118 | 51660 | Hygrocybe glutinipes | R. Haller Aar., 1956 (J.E. Lange)
3070 | 20922 | Hypericum dubium | Leers
8109 | 51651 | Hypericum dubium | Leers
7957 | 51498 | Lactarius vellereus | Fr., 1838 (Fr.)
7248 | 50556 | Lactarius vellereus | (Fr.) Fr. 1838
2995 | 20822 | Lepidium draba | L.
8082 | 51624 | Lepidium draba | L.
8286 | 51854 | Lepraria lobificans | Nyl.
7269 | 50577 | Lepraria lobificans | Nyl.
8254 | 51821 | Lycoperdon pyriforme | Schaeff.
5298 | 30160 | Lycoperdon pyriforme | Schaeff.:Pers.
7736 | 51059 | Malus sylvestris subsp. sylvestris | 
7301 | 50609 | Malus sylvestris subsp. sylvestris | 
875 | 4246 | Morchella esculenta | (L.:Fr.) Pers.
8151 | 51694 | Morchella esculenta | Pers. (L.)
5337 | 30529 | Myxarium nucleatum | Wallr.
7900 | 51441 | Myxarium nucleatum | Wallr., 1833
7056 | 50676 | Oulema | 
6039 | 37486 | Oulema | Goz., 1886
5400 | 30875 | Panellus serotinus | (Pers.) Kühner
7935 | 51476 | Panellus serotinus | Kühner, 1950 (Pers.)
889 | 4269 | Phallus impudicus | L.:Pers.
7824 | 51365 | Phallus impudicus | L., 1753
7426 | 50737 | Phyllonorycter klemannella | (Fabricius, 1781)
3877 | 22507 | Phyllonorycter klemannella | (Fabricius, 1781)
4445 | 24425 | Polistes dominula | (Christ, 1791)
7461 | 50771 | Polistes dominula | (Christ, 1791)
7979 | 51520 | Pseudepidalea viridis | (Laurenti, 1768)
8251 | 51794 | Pseudepidalea viridis | (Laurenti, 1768)
7860 | 51401 | Securigera varia | Lassen (L.)
3145 | 21036 | Securigera varia | (L.) Lassen
3506 | 21519 | Solanum triflorum | Nutt.
7899 | 51440 | Solanum triflorum | Nutt.
7734 | 51057 | Stellaria media subsp. media | 
7571 | 50881 | Stellaria media subsp. media | (L.) Vill.
7928 | 51469 | Stigmatogaster subterranea | (Shaw, 1789)
1254 | 5762 | Stigmatogaster subterranea | (Shaw, 1789)
7849 | 51390 | Tubaria furfuracea | Gillet, 1876 (Pers.)
7627 | 50937 | Tubaria furfuracea | (Pers.) Gillet 1876
8246 | 51789 | Ulmus minor | Mill.
3515 | 21531 | Ulmus minor | Mill.
8292 | 51860 | Xanthoria elegans | Th.Fr. (Link)
5811 | 34181 | Xanthoria elegans | (Link) Th. Fr.

**Comment by damianooldoni** [comment]
### Check result

Good news: I checked all these names and I found a 100% match. So we can safely solve them.

### How solve the duplicates

I would detect duplicates after populating scientificname table and store the mapping in a text file in a tabular way. 
 Example:

id | deprecatedTaxonId | newDeprecatedTaxonId | scientificName | authorhsip
-- | -- | -- | -- | --
4859 | 26718 | 26718 | Agaricus sylvicola | (Vittad.) Peck, 1872
6818 | 50129 | 26718 | Agaricus sylvicola | 
7743 | 51280 | 16405 | Agonum emarginatum | (Gyllenhal, 1827)
2169 | 16405 | 16405 | Agonum emarginatum | (Gyllenhal, 1827)
... | ... | ... | ... | ...

where the column `newDeprecatedTaxonId` is manually set up. We could just take e.g. the lower `deprecatedTaxonId` and fill `newDeprecatedTaxonId` automatically but the number of taxa is not big and this code will be used only once for migration. I would not waste too much time in authomatically assigning the new deprecated taxon id.

So, once we have this file, I would do the following steps:
1. See if new duplicates occur and in case add them to the file for checking ( in our case it will not be the case as we work on a static copy, but we should take care about this when we move to BIM infrastructure).
2. Change the taxonId in `identifiablespecies` based on values of `deprecatedTaxonId` and  `newDeprecatedTaxonId`
3. As we change stuff, we should also take care of revert the changes in the very first step of our `transform_db.py` script. Doing this means that the first step is more than just a sql snippet with only `DROP` statements. But I don't find it a big issue, isn't?

@niconoe : what do you think about it? Do you have something simpler in mind?


So, names in `scientificname` sharing the same `taxonomyId` are deleted. We take for example the name with the lower `id`. In order to not breaking any link with ìdenitifiablespecies` we should save a file where we map the deprecatedTaxonid we are going to replace in identifiablespecies:

**Comment by niconoe** [comment]
Thanks @damianooldoni for pointing me to this issue again! You have a deeper understanding of all this than me, here are however a few impressions (that's mostly notes to clarify my thoughts, let's discuss that in our meeting today!)

- I find the suggested approach a bit convoluted, and would prefer a simpler solution for this problem (if possible!)
- Involving a manually filled tabular file make the process semi-automatic rather than fully automatic (not a blocking point, but to avoid if possible)
- I think we should list the different reasons why those duplicates happen in the first place, and if possible solve at the source rather than later in the process. For example, it seems one of the reasons is that they are obvious duplicates in the existing `taxon` table. I'd solve that first (either manually in their DB before running our script, or by adding a "preparation" step in the script that remove those duplicates before doing anything else)
- Are there other reasons? Can we also solve them at the source?
- Can we add some constraints in our table to ensure that we never enter duplicates in our new tables (defensive programming: we make sure our script work as intended now, but we also prevent more data errors later on)

**Comment by damianooldoni** [comment]
Thanks @niconoe! Actually you got more than a point! :superhero_man: It would be way better to solve these problems beforehand. I agree also about adding a constraint in the making of our tables as this should prevent such issues in the future, making us and BIM aware of them and thus solving them. Again, a nice example of defensive programming.

**Comment by niconoe** [comment]
Post meeting update: I'll work on this: first step is a (separate than transform_db?) script to deduplicate the existing taxon table (and its incoming FK). To solve:

- what about (script execution) repeatability?
- what about manual intervention/configuration (human decisions needed?)

Then, we'll see if there are still duplicates. If so, understand the source of errors and solve it (as early as possible in the process).

**Comment by damianooldoni** [comment]
As mentioned by BIM experts in previous meetings, the source of duplicates is very likely the occurrence import script. To be tackled in #39

**Comment by damianooldoni** [comment]
Maybe it is obvious to you, @niconoe : but I would like to stress what I intend as duplicates and what it is not in `scientificname` table.

A taxon in `scientificname` table is a **duplicate** of another  taxon if it shares the same value of `scientificName` AND the same 
`authorship`.

E.g. this is NOT a duplicate:

id | deprecatedTaxonId | newDeprecatedTaxonId | scientificName | authorhsip
-- | -- | -- | -- | --
4859 | 26718 | 26718 | Agaricus sylvicola | (Vittad.) Peck, 1872
6818 | 50129 | 26718 | Agaricus sylvicola |

The reason is that scientificname is "just" a list of names, the real taxonomy is stored in another table. We don't have to care too musch about the cleanness of this table. Also, trying to have unique values of `scientificName` (without taking into account `authorship`) is conceptually wrong as you can always have cases of _hemihomonyms_ (as I learned from biologists), taxa with same scientific name but belonging to two different kingdoms and where the authorship is crucial to distinguishing them.

Moreover, I would not focus now on solving the kind of issues as shown in table below: these are not duplicates db speaking. They could be improved, yes, but it is definitely not essential for our workflow as both would be linked to the same taxon in `taxonomy` table, so the only problem here is just a call extra to the GBIF API when matching names to GBIF BAckbone, just that:

id | deprecatedTaxonId | newDeprecatedTaxonId | scientificName | authorhsip
-- | -- | -- | -- | --
8285 | 51853 | Caloplaca aurantia | Hellb. (Pers.)
5796 | 34158 | Caloplaca aurantia | (Pers.) Hellbom
